### PR TITLE
[HLSL] AST support for WaveSize attribute.

### PR DESCRIPTION
**Pull request by python3kgae**
First step for support WaveSize attribute in
 https://microsoft.github.io/DirectX-Specs/d3d/HLSL_SM_6_6_WaveSize.html
and https://microsoft.github.io/hlsl-specs/proposals/0013-wave-size-range.html

A new attribute HLSLWaveSizeAttr was supported in the AST.

Implement both the wave size and the wave size range, rather than separately which might require more work.

For #70118

### CODE REVIEW

File: clang/lib/Sema/SemaHLSL.cpp [hunk 1]
```diff
@@ -245,6 +274,20 @@ void SemaHLSL::CheckEntryPoint(FunctionDecl *FD) {
           << llvm::Triple::getEnvironmentTypeName(ST);
       FD->setInvalidDecl();
     }
+    if (const auto *NT = FD->getAttr<HLSLWaveSizeAttr>()) {
+      if (Ver.getMajor() < 6u ||
+          (Ver.getMajor() == 6u && Ver.getMinor() < 6u)) {
```
Comment (bogner) [inline clang/lib/Sema/SemaHLSL.cpp:279]: I think this can be written as
```c++
          if (Ver < VersionTuple(6, 6)) {
```

File: clang/lib/Sema/SemaHLSL.cpp [hunk 2]
```diff
@@ -245,6 +274,20 @@ void SemaHLSL::CheckEntryPoint(FunctionDecl *FD) {
...
+        Diag(NT->getLocation(), diag::err_hlsl_attribute_in_wrong_shader_model)
+            << "wavesize"
+            << "6.6";
+        FD->setInvalidDecl();
+      } else if (NT->getSpelledArgsCount() > 1 &&
+                 (Ver.getMajor() == 6u && Ver.getMinor() < 8u)) {
```
Comment (bogner) [inline clang/lib/Sema/SemaHLSL.cpp:285]: same here

File: clang/lib/Sema/SemaHLSL.cpp [hunk 3]
```diff
@@ -348,6 +391,77 @@ void SemaHLSL::handleNumThreadsAttr(Decl *D, const ParsedAttr &AL) {
...
+  if (SpelledArgsCount > 1 &&
+      !SemaRef.checkUInt32Argument(AL, AL.getArgAsExpr(1), Max))
+    return;
+
+  uint32_t Preferred = 0;
+  if (SpelledArgsCount > 2 &&
+      !SemaRef.checkUInt32Argument(AL, AL.getArgAsExpr(2), Preferred))
+    return;
+  llvm::hlsl::WaveSize WaveSize(Min, Max, Preferred);
+  llvm::hlsl::WaveSize::ValidationResult ValidationResult = WaveSize.validate();
```
Comment (bogner) [inline clang/lib/Sema/SemaHLSL.cpp:415]: What do the `WaveSize.validate()` function and the `ValidationResult` enum gain us here? The logic below pretty much repeats everything that validate did, just checking enums that are named after the various conditions rather than the conditions themselves. Would it be simpler to just implement the logic here and keep it out of the WaveSize object entirely?

Comment (python3kgae) [reply clang/lib/Sema/SemaHLSL.cpp:415]: We'll need to validate WaveSize at llvm IR level as well.
WaveSize::validate could be shared between Sema and DirectX backend.

File: clang/lib/Sema/SemaHLSL.cpp (hunk 1 above)
Comment (python3kgae) [reply clang/lib/Sema/SemaHLSL.cpp]: Updated.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 4]
```diff
@@ -12361,6 +12361,21 @@ def warn_hlsl_availability_unavailable :
 def err_hlsl_export_not_on_function : Error<
   "export declaration can only be used on functions">;
 
+def err_hlsl_attribute_in_wrong_shader_model: Error<
+  "attribute %0 requires shader model %1 or greater">;
+
+def err_hlsl_wavesize_size: Error<
+  "wavesize arguments must be between 4 and 128 and a power of 2">;
+def err_hlsl_wavesize_min_geq_max: Error<
+  "minimum wavesize value %0 must be less than maximum wavesize value %1">;
```
Comment (llvm-beanz) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:12370]: How about `err_attribute_argument_invalid`?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 5]
```diff
@@ -12361,6 +12361,21 @@ def warn_hlsl_availability_unavailable :
...
+def warn_hlsl_wavesize_min_eq_max:  Warning<
+  "wave size range minimum and maximum are equal">,
```
Comment (llvm-beanz) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:12372]: Can we make this more generic and match other attribute formats?
```suggestion
def warn_attr_min_eq_max:  Warning<
  "%0 attribute minimum and maximum arguments are equal">,
```

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 6]
```diff
@@ -12361,6 +12361,21 @@ def warn_hlsl_availability_unavailable :
...
+  InGroup<HLSLAttributeStatement>, DefaultError;
+def err_hlsl_wavesize_pref_size_out_of_range: Error<
+  "preferred wavesize value %0 must be between %1 and %2">;
+def err_hlsl_wavesize_insufficient_shader_model: Error<
+  "wavesize only takes multiple arguments in shader model 6.8 or higher">;
```
Comment (llvm-beanz) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:12377]: Maybe this should just use `err_attribute_wrong_number_arguments`, and we should have a more generic note to provide context that SM 6.8 adds support for the 3-argument variant.

File: clang/lib/Sema/SemaDecl.cpp [hunk 7]
```diff
@@ -2862,6 +2862,10 @@ static bool mergeDeclAttribute(Sema &S, NamedDecl *D,
   else if (const auto *NT = dyn_cast<HLSLNumThreadsAttr>(Attr))
     NewAttr = S.HLSL().mergeNumThreadsAttr(D, *NT, NT->getX(), NT->getY(),
                                            NT->getZ());
+  else if (const auto *NT = dyn_cast<HLSLWaveSizeAttr>(Attr))
```
Comment (llvm-beanz) [inline clang/lib/Sema/SemaDecl.cpp:2865]: This code block looks ripe for a refactoring to use a switch to detect the attribute type... We should consider doing that cleanup in a follow-up change.

File: clang/lib/Sema/SemaHLSL.cpp [hunk 8]
```diff
@@ -245,6 +274,18 @@ void SemaHLSL::CheckEntryPoint(FunctionDecl *FD) {
           << llvm::Triple::getEnvironmentTypeName(ST);
       FD->setInvalidDecl();
     }
+    if (const auto *NT = FD->getAttr<HLSLWaveSizeAttr>()) {
+      if (Ver < VersionTuple(6, 6)) {
+        Diag(NT->getLocation(), diag::err_hlsl_attribute_in_wrong_shader_model)
+            << "wavesize"
```
Comment (llvm-beanz) [inline clang/lib/Sema/SemaHLSL.cpp:280]: You should be able to put the attribute declaration here rather than the string.

File: clang/lib/Sema/SemaHLSL.cpp [hunk 9]
```diff
@@ -245,6 +274,18 @@ void SemaHLSL::CheckEntryPoint(FunctionDecl *FD) {
           << llvm::Triple::getEnvironmentTypeName(ST);
       FD->setInvalidDecl();
     }
+    if (const auto *NT = FD->getAttr<HLSLWaveSizeAttr>()) {
```
Comment (llvm-beanz) [inline clang/lib/Sema/SemaHLSL.cpp:277]: Why is NT the name here? That doesn't seem to make sense with the context of what you're assigning.

File: clang/lib/Sema/SemaHLSL.cpp (hunk 3 above)
Comment (llvm-beanz) [reply clang/lib/Sema/SemaHLSL.cpp:413]: This validation code looks way too complicated and way too hard to follow. I don't think the code sharing here gets us anything and it seems to come at a cost of a lot of extra complexity.

Comment (python3kgae) [reply clang/lib/Sema/SemaHLSL.cpp]: Updated.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 10]
```diff
@@ -12361,6 +12361,21 @@ def warn_hlsl_availability_unavailable :
 def err_hlsl_export_not_on_function : Error<
   "export declaration can only be used on functions">;
 
+def err_hlsl_attribute_in_wrong_shader_model: Error<
+  "attribute %0 requires shader model %1 or greater">;
+
+def err_hlsl_wavesize_size: Error<
+  "wavesize arguments must be between 4 and 128 and a power of 2">;
```
[the remark this answers is not in the document]
Comment (python3kgae) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Updated.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 11]
```diff
@@ -12361,6 +12361,21 @@ def warn_hlsl_availability_unavailable :
...
+def err_hlsl_wavesize_min_geq_max: Error<
+  "minimum wavesize value %0 must be less than maximum wavesize value %1">;
+def warn_hlsl_wavesize_min_eq_max:  Warning<
+  "wave size range minimum and maximum are equal">,
+  InGroup<HLSLAttributeStatement>, DefaultError;
+def err_hlsl_wavesize_pref_size_out_of_range: Error<
+  "preferred wavesize value %0 must be between %1 and %2">;
```
[the remark this answers is not in the document]
Comment (python3kgae) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Updated.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 4 above)
Comment (python3kgae) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Updated

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 5 above)
Comment (python3kgae) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Updated.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 6 above)
Comment (python3kgae) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Updated to 
```
def err_hlsl_attribute_number_arguments_insufficient_shader_model: Error<
  "attribute %0 with %1 arguments requires shader model %2 or greater">;

```

File: clang/lib/Sema/SemaHLSL.cpp (hunk 8 above)
Comment (python3kgae) [reply clang/lib/Sema/SemaHLSL.cpp]: Fixed.

File: clang/lib/Sema/SemaHLSL.cpp (hunk 9 above)
Comment (python3kgae) [reply clang/lib/Sema/SemaHLSL.cpp]: Renamed.

File: clang/test/AST/HLSL/WaveSize.hlsl [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+ [numthreads(8,8,1)]
+ [WaveSize(8, 64)]
+ void w1() {
+ }
+
+
+// CHECK-LABLE:FunctionDecl 0x{{[0-9a-f]+}} <{{.*}}> w2 'void ()'
+// CHECK:HLSLWaveSizeAttr 0x{{[0-9a-f]+}} <{{.*}}> 8 128 64
+ [numthreads(8,8,1)]
+ [WaveSize(8, 128, 64)]
```
Comment (bob80905) [inline clang/test/AST/HLSL/WaveSize.hlsl:23]: Would you consider adding an AST test with multiple wavesize attributes?

**Review by bob80905 — COMMENTED** [review]
Other than my nits, looks good to me

File: clang/test/AST/HLSL/WaveSize.hlsl (hunk 12 above)
Comment (python3kgae) [reply clang/test/AST/HLSL/WaveSize.hlsl:25]: Updated.

File: clang/lib/Sema/SemaDecl.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (python3kgae) [reply clang/lib/Sema/SemaDecl.cpp]: Updated.

File: clang/lib/Sema/SemaHLSL.cpp [hunk 13]
```diff
@@ -144,6 +145,25 @@ HLSLNumThreadsAttr *SemaHLSL::mergeNumThreadsAttr(Decl *D,
...
+                                              int Min, int Max, int Preferred,
+                                              int SpelledArgsCount) {
+  if (HLSLWaveSizeAttr *NT = D->getAttr<HLSLWaveSizeAttr>()) {
+    if (NT->getMin() != Min || NT->getMax() != Max ||
+        NT->getPreferred() != Preferred ||
+        NT->getSpelledArgsCount() != SpelledArgsCount) {
+      Diag(NT->getLocation(), diag::err_hlsl_attribute_param_mismatch) << AL;
+      Diag(AL.getLoc(), diag::note_conflicting_attribute);
+    }
+    return nullptr;
```
[the remark this answers is not in the document]
Comment (python3kgae) [reply clang/lib/Sema/SemaHLSL.cpp:159]: It only ignores when the value matches.
If not match, error will be reported.

**Review by bob80905 — APPROVED** [review]

File: clang/test/SemaHLSL/WaveSize-invalid-profiles.hlsl [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+#elif __SHADER_TARGET_STAGE == __SHADER_STAGE_VERTEX
+// expected-error@+8 {{attribute 'WaveSize' is unsupported in 'vertex' shaders, requires one of the following: compute, amplification, mesh}}
+#elif __SHADER_TARGET_STAGE == __SHADER_STAGE_GEOMETRY
+// expected-error@+6 {{attribute 'WaveSize' is unsupported in 'geometry' shaders, requires one of the following: compute, amplification, mesh}}
+#elif __SHADER_TARGET_STAGE == __SHADER_STAGE_HULL
+// expected-error@+4 {{attribute 'WaveSize' is unsupported in 'hull' shaders, requires one of the following: compute, amplification, mesh}}
+#elif __SHADER_TARGET_STAGE == __SHADER_STAGE_DOMAIN
+// expected-error@+2 {{attribute 'WaveSize' is unsupported in 'domain' shaders, requires one of the following: compute, amplification, mesh}}
+#endif
+[WaveSize(16)]
```
Comment (llvm-beanz) [inline clang/test/SemaHLSL/WaveSize-invalid-profiles.hlsl:18]: If you put something like `// #WaveSize` at the end of the line here, the checks above can all be `@#WaveSize` instead of line offsets.

File: clang/lib/Sema/SemaHLSL.cpp [hunk 15]
```diff
@@ -357,6 +398,74 @@ void SemaHLSL::handleNumThreadsAttr(Decl *D, const ParsedAttr &AL) {
     D->addAttr(NewAttr);
 }
 
+static bool isValidWaveSizeValue(unsigned Value) {
+  return (Value >= 4 && Value <= 128 && ((Value & (Value - 1)) == 0));
```
Comment (llvm-beanz) [inline clang/lib/Sema/SemaHLSL.cpp:402]: Can we instead use `llvm::isPowerOf2_32`? That is clearer to read.

**Review by llvm-beanz — APPROVED** [review]
A few minor suggestions, but otherwise looks good.

File: clang/lib/Sema/SemaHLSL.cpp [hunk 16]
```diff
@@ -153,6 +154,25 @@ HLSLNumThreadsAttr *SemaHLSL::mergeNumThreadsAttr(Decl *D,
       HLSLNumThreadsAttr(getASTContext(), AL, X, Y, Z);
 }
 
+HLSLWaveSizeAttr *SemaHLSL::mergeWaveSizeAttr(Decl *D,
+                                              const AttributeCommonInfo &AL,
+                                              int Min, int Max, int Preferred,
+                                              int SpelledArgsCount) {
+  if (HLSLWaveSizeAttr *NT = D->getAttr<HLSLWaveSizeAttr>()) {
```
Comment (bogner) [inline clang/lib/Sema/SemaHLSL.cpp:161]: One more "NT" to rename.

**Review by bogner — APPROVED** [review]

File: clang/test/SemaHLSL/WaveSize-invalid-profiles.hlsl (hunk 14 above)
Comment (python3kgae) [reply clang/test/SemaHLSL/WaveSize-invalid-profiles.hlsl]: Updated.

File: clang/lib/Sema/SemaHLSL.cpp (hunk 15 above)
Comment (python3kgae) [reply clang/lib/Sema/SemaHLSL.cpp]: Updated.

File: clang/lib/Sema/SemaHLSL.cpp (hunk 16 above)
Comment (python3kgae) [reply clang/lib/Sema/SemaHLSL.cpp]: Fixed.
